Fix icon-request race in attachment ref tests

squelart opened the pull request:
#### 6e502e9d0156d9c2694057afdfb478bc258729de
<pre>
Fix icon-request race in attachment ref tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=258865">https://bugs.webkit.org/show_bug.cgi?id=258865</a>
rdar://problem/111764330

Reviewed by Aditya Keerthi.

Because icons are fetched asynchronously during the first paint, ref-tests should wait until all
Attachments have received their icons, in order to compare fully-rendered attachments.

* LayoutTests/fast/attachment/cocoa/wide-attachment-class-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-class.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-default-icon-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-default-icon.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-folder-icon-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-folder-icon.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-icon-from-file-extension-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-icon-from-file-extension.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-id-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-id.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-uti-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-uti.html:
* LayoutTests/fast/attachment/mac/wide-attachment-element-gpu-process-expected.html:
* LayoutTests/fast/attachment/mac/wide-attachment-element-gpu-process.html:
* LayoutTests/fast/attachment/mac/wide-attachment-title-with-rtl-expected.html:
* LayoutTests/fast/attachment/mac/wide-attachment-title-with-rtl.html:
* LayoutTests/fast/attachment/mac/wide-attachment-type-attribute-expected.html:
* LayoutTests/fast/attachment/mac/wide-attachment-type-attribute.html:
Add reftest-wait&apos;s, and calls to takeScreenshotWhenAttachmentsSettled.

* LayoutTests/fast/attachment/resources/attachment-test-utils.js: Added.
(takeScreenshotWhenAttachmentsSettled):
(takeActualScreenshotWhenAttachmentsSettled):
(takeExpectedScreenshotWhenAttachmentsSettled):
This will wait for all attachments to have received their icon (&quot;load&quot; event) and let the ref-test continue.
Otherwise, add diagnostic next to the failed attachment(s).

* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
Removing recent flakiness flags, as these tests should now always pass.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::HTMLAttachmentElement::updateImage):
(WebCore::HTMLAttachmentElement::updateIconForNarrowLayout):
(WebCore::HTMLAttachmentElement::updateIconForWideLayout):
(WebCore::HTMLAttachmentElement::requestWideLayoutIconIfNeeded):
These functions now update the m_requestingIcon flag as appropriate.
Also the updateIcon function are able to receive an empty icon (just to indicate that we&apos;re received a response).
`&lt;attachment&gt;` now dispatches the following events:
- &quot;beforeload&quot;: There is a need for an icon, a request has been made.
- &quot;loadingerror&quot;: Request response didn&apos;t contain an image.
- &quot;loadeddata&quot;: Image from the request response has been set into the inner `&lt;img&gt;`.
- &quot;error&quot;: Error from `&lt;img&gt;`.
- &quot;load&quot;: Success from `&lt;img&gt;`, indicated the icon should now be visible.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateAttachmentIcon):
In case there is no actual image in the response, inform Attachment by sending null/empty data.

Canonical link: <a href="https://commits.webkit.org/266144@main">https://commits.webkit.org/266144@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/306b9d8a0a3e63138f682577f09d95c98fb7697c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12972 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13292 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13623 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14710 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12376 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13040 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15797 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13317 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15073 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13138 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13838 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10974 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15161 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11126 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11729 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18792 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12201 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11895 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15106 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12366 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10256 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11629 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11595 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3191 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15947 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12207 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/attachment/resources/attachmentTestUtils.js:
nit: we usually name it `attachment-test-utils.js` or `attachment-helper.js`

squelart replied:
I saw camelCase elsewhere. But ok, will change.

nt1m commented on LayoutTests/fast/attachment/resources/attachmentTestUtils.js:
```diff
@@ -0,0 +1,20 @@
+
+    window.addEventListener("load", async function() {
+        let attachments = Array.from(document.getElementsByTagName("attachment"));
+        for (let i = 0; i < 10 + attachments.length; ++i) {
```
Why do we stop after an arbitrary limit? Shouldn't we keep polling until the condition (`allIconsLoaded`) is met?

squelart replied:
It's a kind of timeout. In practice icons should be there in 2-4 iterations, so if by 10 it's not there yet, just give up.
Would you prefer an infinite loop, and rely on the test overall timeout?

squelart replied:
Ok, made it an infinite loop. I guess it could catch unexpected never-getting-an-icon cases.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.idl:
```diff
@@ -35,4 +35,6 @@
 
     static DOMString getAttachmentIdentifier(HTMLImageElement imageElement);
+
+    readonly attribute boolean hasThumbnailOrIcon;
```
This information should probably be exposed via `Internals.idl`, since it doesn't need to be a part of the client interface.

nt1m commented on LayoutTests/fast/attachment/resources/attachment-test-utils.js:
```diff
@@ -0,0 +1,20 @@
+
+        document.documentElement.classList.remove("reftest-wait");
+    });
+}
```
```suggestion
async function takeScreenshotWhenAttachmentsSettled()
{
    console.assert(document.documentElement.classList.contains("reftest-wait"));
    if (document.readyState != "complete")
        await new Promise(() => addEventListener("load", takeScreenshotWhenAttachmentsSettled, { once: true }));
    let attachments = [...document.getElementsByTagName("attachment")];
    console.assert(attachments.length);
    if (attachments.every(a => a.hasThumbnailOrIcon))
        document.documentElement.classList.remove("reftest-wait");
    else
        requestAnimationFrame(takeScreenshotWhenIconLoaded);
}
```
You can write this recursively if you want.

squelart replied:
I've changed to your style, thanks.

However I've re-added an iteration count, as I found a few tests that timed out. I'll try to debug, but it seems very difficult to get any logging or to attach-to&debug these layout tests!

squelart replied:
I haven't found the cause of the timeouts, but now if the iteration count goes too high, a comment is inserted in the web page, which makes it fail (as it would have anyway, but quicker) and points at the issue.

pxlcoder commented on LayoutTests/fast/attachment/mac/wide-attachment-element-gpu-process-expected.html:
```diff
@@ -3,4 +3,4 @@
 </body>
-</html>
\ No newline at end of file
+</html>
```
Unintentional change?

squelart replied:
Actually intentional, I wanted to get rid of the "no newline" message while I was nearby. Should I undo?

pxlcoder replied:
Fine to keep.

pxlcoder commented on LayoutTests/fast/attachment/mac/wide-attachment-element-gpu-process.html:
```diff
@@ -6,4 +6,4 @@
 </body>
-</html>
\ No newline at end of file
+</html>
```
Unintentional change?

squelart replied:
Actually intentional, I wanted to get rid of the "no newline" message while I was nearby. Should I undo?

pxlcoder commented on LayoutTests/fast/attachment/resources/attachment-test-utils.js:
```diff
@@ -0,0 +1,46 @@
+function hasThumbnailOrIcon(elem)
```
`elem` -> `element`, we try to avoid shorthand variable names in WebKit.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8024,21 +8024,30 @@ void WebPage::updateAttachmentIcon(const String& identifier, ShareableBitmap::Ha
         if (File* file = attachment->file())
             fileName = file->path();
-        send(Messages::WebPageProxy::RequestAttachmentIcon(identifier, attachment->attachmentType(), fileName, attachment->attachmentTitle(), size));
+        return send(Messages::WebPageProxy::RequestAttachmentIcon(identifier, attachment->attachmentType(), fileName, attachment->attachmentTitle(), size));
```
I don't think we normally use the return value of `send`. I would just return `true` after calling it.

squelart replied:
Well, the goal was to know if the message was actually sent, so the caller `HTMLAttachmentElement::requestWideLayoutIconIfNeeded()` knows whether to set its `m_requestingIcon`: If the message was not even sent, then we're not waiting for an icon.

pxlcoder replied:
I guess it's fine. Just not something we normally do, as something has gone terribly wrong if `send` fails.

squelart replied:
I've now removed that code.

pxlcoder commented on LayoutTests/fast/attachment/resources/attachment-test-utils.js:
```diff
@@ -0,0 +1,46 @@
+        return document.documentElement.classList.remove("reftest-wait");
+    }
+
+    requestAnimationFrame(() => takeScreenshotWhenAttachmentsSettled(message, iteration + 1));
```
How difficult do you think it would be to make this event-based, rather than polling for the icon?

squelart replied:
Good idea, I'll have a look...

squelart replied:
What did you make me do? 😄
I've removed the polling code and functions, now it's all event-driven.

nt1m commented on LayoutTests/fast/attachment/cocoa/wide-attachment-class-expected.html:
```diff
@@ -1,7 +1,11 @@
 <!DOCTYPE html><!-- webkit-test-runner [ AttachmentElementEnabled=true AttachmentWideLayoutEnabled=true ] -->
-<html>
+<html class='reftest-wait'>
```
nit: double quotes across the patch for consistency

nt1m commented on LayoutTests/fast/attachment/resources/attachment-test-utils.js:
```diff
@@ -0,0 +1,51 @@
+function takeScreenshotWhenAttachmentsSettled(message, iteration = 0)
```
iteration is no longer used

nt1m commented on LayoutTests/fast/attachment/resources/attachment-test-utils.js:
```diff
@@ -0,0 +1,51 @@
+                });
+            }
+        });
+    }
```
I would write this whole function as:

```js
function takeScreenshotWhenAttachmentsSettled() {
    console.assert(document.documentElement.classList.contains("reftest-wait"));

    const attachments = [...document.getElementsByTagName("attachment")];
    console.assert(attachments.length);
    if (!attachments.length)
        return document.documentElement.classList.remove("reftest-wait");

    const states = new Map(attachments.map(a => [a, []]));

    const onFailure = () => {
        for (const [attachment, events] of states)
            attachment.insertAdjacentText("afterend", "<- (" + message + ") - events = [" + events.join() + "]");
        document.documentElement.classList.remove("reftest-wait");
    };

    const timeoutId = setTimeout(onFailure, 5000);

    const onSuccess = () => {
        clearTimeout(timeoutId);
        document.documentElement.classList.remove("reftest-wait");
    };

    const promises = attachments.map((attachment) => new Promise((resolve, reject) => {
        for (const event of ["beforeload", "loadingerror", "loadeddata", "error"])
            attachment.addEventListener(event, () => { states.get(attachment).push(event); });
        attachment.addEventListener("loadingerror", reject, { once: true });
        attachment.addEventListener("load", () => {
            requestAnimationFrame(() => requestAnimationFrame(resolve));
        }, { once: true });
    }));

    Promise.all(promises).then(onSuccess, onFailure);
}
```

(You don't need the double rAF on failure, and generally this looks cleaner using promises)
(I also added early exit if we reach loadingerror, instead of waiting for the timeout, but feel free to remove that if you prefer).

squelart replied:
Thank you. The only change I've added, is to also push "load": In case there are multiple attachments, we'll know those that have finished their load.

squelart committed:
Fix icon-request race in attachment ref tests
https://bugs.webkit.org/show_bug.cgi?id=258865
rdar://problem/111764330

Reviewed by Aditya Keerthi.

Because icons are fetched asynchronously during the first paint, ref-tests should wait until all
Attachments have received their icons, in order to compare fully-rendered attachments.

* LayoutTests/fast/attachment/cocoa/wide-attachment-class-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-class.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-default-icon-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-default-icon.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-folder-icon-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-folder-icon.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-icon-from-file-extension-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-icon-from-file-extension.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-id-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-id.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-uti-expected.html:
* LayoutTests/fast/attachment/cocoa/wide-attachment-uti.html:
* LayoutTests/fast/attachment/mac/wide-attachment-element-gpu-process-expected.html:
* LayoutTests/fast/attachment/mac/wide-attachment-element-gpu-process.html:
* LayoutTests/fast/attachment/mac/wide-attachment-title-with-rtl-expected.html:
* LayoutTests/fast/attachment/mac/wide-attachment-title-with-rtl.html:
* LayoutTests/fast/attachment/mac/wide-attachment-type-attribute-expected.html:
* LayoutTests/fast/attachment/mac/wide-attachment-type-attribute.html:
Add reftest-wait's, and calls to takeScreenshotWhenAttachmentsSettled.

* LayoutTests/fast/attachment/resources/attachment-test-utils.js: Added.
(takeScreenshotWhenAttachmentsSettled):
(takeActualScreenshotWhenAttachmentsSettled):
(takeExpectedScreenshotWhenAttachmentsSettled):
This will wait for all attachments to have received their icon ("load" event) and let the ref-test continue.
Otherwise, add diagnostic next to the failed attachment(s).

* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
Removing recent flakiness flags, as these tests should now always pass.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::HTMLAttachmentElement::updateImage):
(WebCore::HTMLAttachmentElement::updateIconForNarrowLayout):
(WebCore::HTMLAttachmentElement::updateIconForWideLayout):
(WebCore::HTMLAttachmentElement::requestWideLayoutIconIfNeeded):
These functions now update the m_requestingIcon flag as appropriate.
Also the updateIcon function are able to receive an empty icon (just to indicate that we're received a response).
`<attachment>` now dispatches the following events:
- "beforeload": There is a need for an icon, a request has been made.
- "loadingerror": Request response didn't contain an image.
- "loadeddata": Image from the request response has been set into the inner `<img>`.
- "error": Error from `<img>`.
- "load": Success from `<img>`, indicated the icon should now be visible.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateAttachmentIcon):
In case there is no actual image in the response, inform Attachment by sending null/empty data.

Canonical link: https://commits.webkit.org/266144@main
